clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -208,11 +208,13 @@ void run(long now) {
             }
         }
 
-        List<RecordBatch> expiredBatches = this.accumulator.abortExpiredBatches(this.requestTimeout, now);
-        // update sensors
-        for (RecordBatch expiredBatch : expiredBatches)
-            this.sensors.recordErrors(expiredBatch.topicPartition.topic(), expiredBatch.recordCount);
-
+        if (!metadata.hasExpired(now)) {
```

Comment:
This logic should probably be put in `RecordAccumulator.abortExpiredBatches`. Otherwise, if we put this logic here and we call RecordAccumulator.abortExpiredBatches() from another place, RecordAccumulator.abortExpiredBatches() would still return partition to be expired even if metadata has expired.